python/private/pypi/uv_pip_compile.bzl

```diff
@@ -0,0 +1,101 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Rule for locking third-party dependencies with uv."
+
+script_template = """\
+{uv} pip compile \
+--python {python} \
+--python-platform windows \
+--python-version 3.9 \
+--no-strip-extras \
+--generate-hashes \
+--output-file - \
+{dependencies_file}
+"""
+
+def _uv_pip_compile(ctx):
+    info = ctx.toolchains["//python:uv_toolchain_type"].uvtoolchaininfo
+    uv = info.binary
+
+    python = ctx.toolchains["@bazel_tools//tools/python:toolchain_type"].py3_runtime.interpreter
+    dependencies_file = ctx.file.dependencies_file
+
+    # Option 1: Build action option.
+    # Not really appropriate, but it executes.
+
+    # args = ctx.actions.args()
+    # args.add("pip")
+    # args.add("compile")
+
+    # uv will use this python for operations where it needs to execute python code. See: UV_PYTHON and https://github.com/astral-sh/uv?tab=readme-ov-file#installing-into-arbitrary-python-environments
+    # args.add("--python", python)
+    # args.add("--python-platform", "windows")
+    # args.add("--python-version", "3.9")
+    # args.add("--no-strip-extras")
+    # args.add("--generate-hashes")
+    #requirements_out = ctx.actions.declare_file(ctx.label.name + ".requirements.out")
+    # args.add("--output-file", requirements_out)
+    # args.add(dependencies_file)
+
+    # ctx.actions.run(
+    #     executable = uv,
+    #     arguments = [args],
+    #     inputs = [dependencies_file],
+    #     outputs = [requirements_out],
+    #     tools = [python],
+    # )
+
+    # Option 2: Run action option.
+    # Works to exec uv --version, but will need rest of arguments plumbed through.
+
+    # executable = ctx.actions.declare_file("%s-uv" % ctx.label.name)
+    # ctx.actions.symlink(
+    #     is_executable = True,
+    #     output = executable,
+    #     target_file = uv,
+    # )
+
+    # Option 3: Run action option.
+    # Works to exec uv with some (but not all) arguments plumbed through. Output to a directory of the resolved output needs to be done.
+    executable = ctx.actions.declare_file("{name}-{uv_name}".format(
+        name = ctx.label.name,
+        uv_name = uv.basename,
+    ))
+    script_content = script_template.format(
+        uv = uv.path,
+        python = python.path,
+        dependencies_file = dependencies_file.path,
+    )
+    ctx.actions.write(executable, script_content, is_executable = True)
+
+    return [DefaultInfo(
+        files = depset([executable]),
+        executable = executable,
+        runfiles = ctx.runfiles([uv, dependencies_file, python]),
+    )]
+
+uv_pip_compile = rule(
+    implementation = _uv_pip_compile,
+    attrs = {
+        "constraints_file": attr.label(allow_single_file = True),
+        "dependencies_file": attr.label(allow_single_file = True),
+        "overrides_file": attr.label(allow_single_file = True),
```

Comment:
What are constraints and overrides? I'm guessing dependencies is the requirements.txt (or equiv) file?

Reply:
+1 for just having `srcs` and `out` in here. This is the common language for all rules.

Reply:
As above, I'm not set on the interface for this rule. These are just some pieces hanging around.

Though, since you're curious, in my vision for this functionality, I would prefer to support the PEP-621 pyproject.toml file (and maybe optionally requirements.in but I think it's really old at this point).

I'd also like support for `constraints` files. Which are requirements specifiers which guide the resolver, but do not need to appear in the lockfile output (this can be useful to stop long backtracking).

I'd also like to support `overrides` files. Which are a useful way to ignore bad requirement specifiers from packages. So it lets you say to the resolver: "I don't care what you think, Im confident that `torch==2.2.1` is what I want in my output".

I'd also like to lock dependencies for multiple python platforms and figure out what the best mapping for this to bazel would be:
```
          - x86_64-pc-windows-msvc:     An x86 Windows target
          - x86_64-unknown-linux-gnu:   An x86 Linux target. Equivalent to `x86_64-manylinux_2_17`
          - aarch64-apple-darwin:       An ARM-based macOS target, as seen on Apple Silicon devices
          - x86_64-apple-darwin:        An x86 macOS target
          - aarch64-unknown-linux-gnu:  An ARM64 Linux target. Equivalent to `aarch64-manylinux_2_17`
          - aarch64-unknown-linux-musl: An ARM64 Linux target
          - x86_64-unknown-linux-musl:  An `x86_64` Linux target
          - x86_64-manylinux_2_17:      An `x86_64` target for the `manylinux_2_17` platform
          - x86_64-manylinux_2_28:      An `x86_64` target for the `manylinux_2_28` platform
          - aarch64-manylinux_2_17:     An ARM64 target for the `manylinux_2_17` platform
          - aarch64-manylinux_2_28:     An ARM64 target for the `manylinux_2_28` platform
```
Even if that means running in a loop across the 1-3 platforms I care about. `uv` is still fast enough to do that.

etc.